### PR TITLE
Attempt to add holo-admin to holoportos-shell

### PR DESCRIPTION
**Pull request by pjkundert**
Unsuccessfully trying to get holo-admin package added to the `holoportos-shell` (not there), or `holoportos-shell demo` (infinite recursion).

### CODE REVIEW

**Comment by pjkundert** [comment]
I tried, unsuccessfully, to follow the model of `holo-envoy`, to add the service, and add it to the `holoportos-shell`.  Previously, it was only in the `demo` version, so I tried to add it to that; not quite sure how to add it to the base `holoportos-shell` version.

File: overlays/holo-nixpkgs/default.nix [hunk 1]
```diff
@@ -21,8 +21,8 @@ let
   holo-config = fetchFromGitHub {
     owner = "Holo-Host";
     repo = "holo-config";
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/default.nix]:
```suggestion
    repo = "hpos-state";
```

File: overlays/holo-nixpkgs/default.nix [hunk 2]
```diff
@@ -21,8 +21,8 @@ let
   holo-config = fetchFromGitHub {
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/default.nix]:
```suggestion
  hpos-state = fetchFromGitHub {
```

File: overlays/holo-nixpkgs/default.nix [hunk 3]
```diff
@@ -59,9 +59,9 @@ in
   inherit (callPackage gitignore {}) gitignoreSource;
 
   inherit (callPackage holo-config {})
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/default.nix]:
```suggestion
  inherit (callPackage hpos-state {})
```

File: profiles/holoportos/demo/default.nix [hunk 4]
```diff
@@ -37,6 +37,7 @@ in
   networking.firewall.allowedTCPPorts = [ 1111 2222 3333 8800 8880 8888 48080 ];
 
   services.holo-envoy.enable = true;
+  services.holo-admin.enable = true;
```
Comment (yegortimoshenko) [inline profiles/holoportos/demo/default.nix]:
```suggestion
  services.hpos-admin.enable = true;
```

File: modules/default.nix [hunk 5]
```diff
@@ -6,6 +6,7 @@
   imports = [
     ./boot/generic-extlinux-compatible.nix
     ./services/holo-envoy.nix
+    ./services/holo-admin.nix
```
Comment (yegortimoshenko) [inline modules/default.nix]:
```suggestion
    ./services/hpos-admin.nix
```

File: modules/services/holo-admin.nix [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    environment.systemPackages = with pkgs; [
+      hpos-state-gen-cli # for testing
+    ];
+    systemd.services.holo-admin = {
+      after = [ "network.target" ];
+      path = [ config.services.holochain-conductor.package ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = "${pkgs.python3}/bin/python3 ${cfg.package}/holo-admin.py";
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]:
```suggestion
        ExecStart = "${cfg.package}/bin/hpos-admin";
```

File: modules/services/holo-admin.nix [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+with lib;
+
+let
+  cfg = config.services.holo-admin;
+in
+
+{
+  options.services.holo-admin = {
+    enable = mkEnableOption "Holo Admin";
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]:
```suggestion
    enable = mkEnableOption "HPOS Admin";
```

File: modules/services/holo-admin.nix [hunk 8]
```diff
@@ -0,0 +1,38 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.holo-admin;
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]:
```suggestion
  cfg = config.services.hpos-admin;
```

File: modules/services/holo-admin.nix [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+in
+
+{
+  options.services.holo-admin = {
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]:
```suggestion
  options.services.hpos-admin = {
```

File: modules/services/holo-admin.nix [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    enable = mkEnableOption "Holo Admin";
+
+    package = mkOption {
+      default = pkgs.holo-admin;
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]:
```suggestion
      default = pkgs.hpos-admin;
```

File: modules/services/holo-admin.nix [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    };
+  };
+
+  # TODO: enable 'sudo nixos-rebuild ...' and 'sudo zerotier-cli -j info`
+  # TODO: make "after" the systemd phase that locates the hpos-state.json file
+  # TODO: set the HPOS_STATE_PATH environment variable
+  config = mkIf cfg.enable {
+    environment.systemPackages = with pkgs; [
+      hpos-state-gen-cli # for testing
+    ];
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]: Adding `hpos-state-gen-cli` to environment doesn't belong here. If you need it for testing, move it to `tests/hpos-admin/default.nix` instead.

File: modules/services/holo-admin.nix [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    systemd.services.holo-admin = {
+      after = [ "network.target" ];
+      path = [ config.services.holochain-conductor.package ];
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]:
```suggestion
```

File: modules/services/holo-admin.nix [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = "${pkgs.python3}/bin/python3 ${cfg.package}/holo-admin.py";
+        KillMode = "process";
+        Restart = "always";
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]:
```suggestion
```

File: modules/services/holo-admin.nix [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+      hpos-state-gen-cli # for testing
+    ];
+    systemd.services.holo-admin = {
+      after = [ "network.target" ];
+      path = [ config.services.holochain-conductor.package ];
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        ExecStart = "${pkgs.python3}/bin/python3 ${cfg.package}/holo-admin.py";
+        KillMode = "process";
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]:
```suggestion
```

File: overlays/holo-nixpkgs/default.nix [hunk 15]
```diff
@@ -129,6 +129,7 @@ in
 
   extlinux-conf-builder = callPackage ./extlinux-conf-builder {};
 
+  inherit (callPackage holo-admin {}) holo-admin;
```
Comment (yegortimoshenko) [inline overlays/holo-nixpkgs/default.nix]:
```suggestion
```

File: tests/holo-admin/default.nix [hunk 16]
```diff
@@ -0,0 +1,24 @@
+{ lib, makeTest }:
+
+makeTest {
+  name = "holo-admin";
+
+  machine = {
```
Comment (yegortimoshenko) [inline tests/holo-admin/default.nix]: For example:

```suggestion
  machine = {
    environment.systemPackages = with pkgs; [ hpos-state-gen-cli ];
```

File: tests/holo-admin/default.nix [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    virtualisation.memorySize = 2048;
+  };
+
+  testScript = ''
+    startAll;
+    # TODO: Create deterministic config somewhere == HPOS_STATE_PATH
+    $machine->succeed(
+	"hpos-state-gen-cli --email a@b.ca --password password > /var/lib/holochain-conductor/holo"
+    );
+    $machine->systemctl("start holo-envoy.service");
```
Comment (yegortimoshenko) [inline tests/holo-admin/default.nix]:
```suggestion
    $machine->systemctl("start hpos-admin.service");
```

File: modules/services/holo-admin.nix [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+  };
+
+  # TODO: enable 'sudo nixos-rebuild ...' and 'sudo zerotier-cli -j info`
+  # TODO: make "after" the systemd phase that locates the hpos-state.json file
+  # TODO: set the HPOS_STATE_PATH environment variable
+  config = mkIf cfg.enable {
+    environment.systemPackages = with pkgs; [
+      hpos-state-gen-cli # for testing
+    ];
+    systemd.services.holo-admin = {
```
Comment (yegortimoshenko) [inline modules/services/holo-admin.nix]: Set up something like this for now:

```suggestion
    systemd.services.holo-admin = {
      environment.HPOS_STATE_PATH = "/tmp/hpos-state.json";
```

I'll fix that up once HPOS Init work is done.

File: tests/holo-admin/default.nix [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+  machine = {
+    imports = [ (import ../../profiles/holoportos/demo) ];
+    virtualisation.memorySize = 2048;
+  };
+
+  testScript = ''
+    startAll;
+    # TODO: Create deterministic config somewhere == HPOS_STATE_PATH
+    $machine->succeed(
+	"hpos-state-gen-cli --email a@b.ca --password password > /var/lib/holochain-conductor/holo"
```
Comment (yegortimoshenko) [inline tests/holo-admin/default.nix]:
```suggestion
	"hpos-state-gen-cli --email a@b.ca --password password > /tmp/hpos-state.json"
```

File: tests/holo-admin/default.nix [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+    );
+    $machine->systemctl("start holo-envoy.service");
+    $machine->waitForUnit("holo-admin.service");
```
Comment (yegortimoshenko) [inline tests/holo-admin/default.nix]:
```suggestion
    $machine->waitForUnit("hpos-admin.service");
```

**Pull request merged**